Made django-sitetree thread safe

agronick opened the pull request:
I remove the global singleton and the global context variable. I'm storing the sitetree class in the context for the request. The first  tag that uses the sitetree will create it. The ones after that will reference the already created sitetree. Everything appears to work for me although you probably will want to go over everything yourself.

If you put some debug statements in create_from_context:

``` python
def sitetree_from_context(context): 
    if 'sitetree_obj' in context:
        print 'Already Exists'
        return context['sitetree_obj']
    context['sitetree_obj'] = get_sitetree()
    print 'Creating tree'
    return context['sitetree_obj']
```

It prints out:

```
Creating tree
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
Already Exists
```

This seems to follow the same pattern on instantiating the sitetree once per request without storing anything in global variables.

idlesign commented:
As you can see CI tests failed.
You've changed a lot of sitetree init machinery but failed to provide a single test case showing a problem before and after. 
Sorry, can't accept it this way.

agronick commented:
This isn't a small issue. Take what I wrote or do it yourself. I don't think you understand how big of an issue this is. At the very least you should tell users to run with one thread per process so users don't get someone else's navigation menu. You don't need a test case to explain this:

I init my menu
I start building my menu
Someone else inits a menu
I now use another user as my user, another request as my request and give you half of someone elses menu
Someone else finishes building their menu

idlesign commented:
I don't think you hear what I saying too.
You don't need to explain the issue more than once. 
We do need tests.

agronick commented:
I changed global_context to context but I didn't update the tests. I updated the tests to reflect that. There was one spot in the tests where it was using the context from an old test in the global var. It puts the new context in. On any of the tags (like url) where it wasn't doing an init_tree I had it do an init_tree if the context was not set. I'm only getting 4 errors now. Some of it may have to do with it expecting the variables that were global to persist between tests. A bunch of test are giving this error which makes no sense:  <class 'mock.MagicMock'>: it's not the same object as mock.MagicMock

idlesign commented:
That's something, thank you.
I'll check it out in a week.

agronick commented:
Thanks.

idlesign commented:
Per thread context is implemented in `feat/thread_context`. Current tests are passing, no special thread safety tests provided yet.
If you have time you're welcome to introduce such tests, so that we could run them against both your and mine implementations.

agronick commented:
I'll look at it tomorrow. I dont care what implementation is used. Maybe test it by running two requests. On previous versions if you canged the context you would change it on both instances. Then you can validate the context is per sitetree. I'll see what I can do at work tomorrow.

idlesign commented:
Superseded by #224. Closing.

idlesign commented:
Locked after #224
